packages/formspree-core/src/forms.ts

```diff
@@ -67,22 +67,23 @@ export function isKnownError(error: FormError): error is KnownError<FormError> {
   );
 }
 
-export interface SuccessBody {
-  id: string;
-  data: object;
+export interface SuccessResponse {
+  body: {
+    id: string;
+    data: object;
+  };
+  response: Response | null;
 }
 
-export interface ErrorBody {
-  errors: FormError[];
+export interface ErrorResponse {
+  body: { errors: FormError[] };
+  response: Response | null;
 }
 
-export type SubmissionBody = SuccessBody | ErrorBody;
+export type SubmissionResponse = SuccessResponse | ErrorResponse;
 
-export function hasErrors(body: SubmissionBody): body is ErrorBody {
-  return (body as ErrorBody).errors !== undefined;
-}
-
-export interface SubmissionResponse {
-  body: SubmissionBody;
-  response: Response | null;
+export function hasErrors(
+  response: SubmissionResponse
+): response is ErrorResponse {
+  return (response as ErrorResponse).body.errors !== undefined;
```

Comment:
This is fine.

Alternatively, if we use discriminated union either via a field tag, users don't have to import the type assertion function; I think it's easier to use and we don't have to have this function.

Here is one example:

```typescript
type SubmissionResult<R> = SubmissionSuccess<R> | SubmissionFailed;

type SubmissionSuccess<R> = {
  kind: 'success';
  id: string;
  data: R;
  response: Response | null;
};

type SubmissionFailed = {
  kind: 'failed';
  errors: Error[];
  response: Response | null;
};

function doSomething<R>(result: SubmissionResult<R>) {
  if (result.kind === 'success') {
    return result.data;
  }
  return result.errors;
}
```

Also, I generally try to avoid doing type assertion (`as MyType`) since it can easily hide wrong type

Reply:
Thinking about this some more. I think `hasError` is the good public API so users don't rely on the specific shape of the result object.